### ISSUE TITLE
Fatal: java.lang.RuntimeException encountered on Ubuntu VM through SSH

### ISSUE THREAD

**Issue by YoDaMa** [Microsoft/Git-Credential-Manager-for-Mac-and-Linux#103]
**Operating System:** Ubuntu 18.04.1 LTS (Bionic Beaver)
**Issue:** Credential Manager is not working on first `git pull`. 
**Repro Steps:** 

Using MobaXTerm as an SSH client.
Following installation using homebrew: 
```
$ sudo apt install linuxbrew-wrapper
$ brew install git-credential-manager
$ git-credential-manager install
```

Then on a Git push:

```
$ git push
Fatal: java.lang.RuntimeException encountered.  Details:
Secure credential storage is not available on this operating system. You may opt-in to store credentials in an unencrypted file under your user home directory by running 'git config --global credential.canFallBackToInsecureStore true'.
fatal: credential helper '!/home/linuxbrew/.linuxbrew/Cellar/jdk/9.0.4/bin/java -Ddebug=false -Djava.net.useSystemProxies=true -jar /home/linuxbrew/.linuxbrew/Cellar/git-credential-manager/2.0.4/libexec/git-credential-manager-2.0.4.jar' told us to quit
```

Not sure the cause of this error.

**Comment by mattcobley** [comment]
I get the exact same error on a fresh install or liveCD of Ubuntu 18.04 or Kubuntu 18.04.  The process I am following is the following:
Install Java 8 using the Oracle installer by running:
```
sudo add-apt-repository ppa:webupd8team/java
sudo apt update
sudo apt install oracle-java8-installer
```
Checked, and Java is available in my path - i.e. if I just type "java" the help options appear.

Then installed Linuxbrew following the instructions here: http://linuxbrew.sh/ i.e. running:
```
sudo apt install curl
sh -c "$(curl -fsSL https://raw.githubusercontent.com/Linuxbrew/install/master/install.sh)"

test -d ~/.linuxbrew && PATH="$HOME/.linuxbrew/bin:$HOME/.linuxbrew/sbin:$PATH"
test -d /home/linuxbrew/.linuxbrew && PATH="/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:$PATH"
test -r ~/.bash_profile && echo "export PATH='$(brew --prefix)/bin:$(brew --prefix)/sbin'":'"$PATH"' >>~/.bash_profile
echo "export PATH='$(brew --prefix)/bin:$(brew --prefix)/sbin'":'"$PATH"' >>~/.profile
```

Then installed GCM using the installation instructions in the Install.md file:

```
brew update
brew install git-credential-manager
git-credential-manager install
```

Note that nothing appears to happen after running the last command - not sure if that is okay or not.

Then tried running: 
`git clone <URL from VSTS> `
as usual to clone a repo (which works fine in Git for Windows with Git Credential Manager installed), and get the same error that YoDaMa was getting above when just running that command:

```
Fatal: java.lang.RuntimeException encountered.  Details:
Secure credential storage is not available on this operating system. You may opt-in to store credentials in an unencrypted file under your user home directory by running 'git config --global credential.canFallBackToInsecureStore true'.
fatal: credential helper '!/usr/lib/jvm/java-8-oracle/jre/bin/java -Ddebug=false -Djava.net.useSystemProxies=true -jar /home/ubuntu/.linuxbrew/Cellar/git-credential-manager/2.0.4/libexec/git-credential-manager-2.0.4.jar' told us to quit
```

**Comment by pjdv2** [comment]
> brew update
> brew install git-credential-manager
> git-credential-manager install
> ```
> 
> Note that nothing appears to happen after running the last command - not sure if that is okay or not.
> 

This 'install' updates the Git configuration file in your home folder (as per the manual, but you can verify this by running `git config --list`).

> Then tried running:
> `git clone <URL from VSTS> `
> as usual to clone a repo (which works fine in Git for Windows with Git Credential Manager installed), and get the same error that YoDaMa was getting above when just running that command:
> 
> ```
> Fatal: java.lang.RuntimeException encountered.  Details:
> Secure credential storage is not available on this operating system. You may opt-in to store credentials in an unencrypted file under your user home directory by running 'git config --global credential.canFallBackToInsecureStore true'.
> fatal: credential helper '!/usr/lib/jvm/java-8-oracle/jre/bin/java -Ddebug=false -Djava.net.useSystemProxies=true -jar /home/ubuntu/.linuxbrew/Cellar/git-credential-manager/2.0.4/libexec/git-credential-manager-2.0.4.jar' told us to quit
> ```
I had the exact same problem. I tried many different things, but also keep getting the same error. I decided to use [SSH authentication](https://docs.microsoft.com/en-us/azure/devops/repos/git/use-ssh-keys-to-authenticate?view=vsts) instead. If you change the debug argument to true you get a longer error output when running `git clone` and may be useful for those fixing this issue.

**Comment by mattcobley** [comment]
@pjdv2 thanks for the reminder about SSH authentication as an alternative.  Would still be nice for this to get fixed though.

**Comment by pjdv2** [comment]
> @pjdv2 thanks for the reminder about SSH authentication as an alternative. Would still be nice for this to get fixed though.

Yes, I agree. Ubuntu 18.10 was the first release that I tried using this manager, so I'm not aware of how it functioned in previous versions. When looking at the debug error in #97, the issue seems to go wrong on the `createSecureStore`-method, and the subsequent error that is displayed: `Unable to load library 'gnome-keyring'`. I read a [post](https://stackoverflow.com/questions/1403788/java-lang-unsatisfiedlinkerror-no-dll-in-java-library-path) that speaks about the specific Java exception `UnsatisfiedLinkError' and suggesting to add a parameter to the java executable as to set the Java library path (i.e., `Djava.library.path=/path/to/dir`). This may fix the issue, but the manager should find that library by default. Apparently, on Linux, the library path value maps to `LD_LIBRARY_PATH`. If you find the library (`gnome-keyring`) location and add this to the java executable and [manually configure git](https://github.com/Microsoft/Git-Credential-Manager-for-Mac-and-Linux/blob/master/Install.md), this may at least get you passed the Exception, and could work. Alternatively, you could try to set the library path to find the necessary `gnome-keyring` library.

**Comment by pjdv2** [comment]
> @pjdv2 thanks for the reminder about SSH authentication as an alternative. Would still be nice for this to get fixed though.

I'm not a 'contributor' on this project. I did found out that the `gnome-keyring` needs to be installed (use `sudo apt install libgnome-keyring0` to do so). There are also some requirements given that seem to indicate that _Oracle Java_ (or _OpenJDK_) needs to be installed. You should be able to get pass the exception, however, in my case, I can't seem to authenticate (even though a [PAT](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=vsts) has been created) and I can't see where this is going wrong (this seems to require actual debugging).

**Comment by pjdv2** [comment]
> @pjdv2 thanks for the reminder about SSH authentication as an alternative. Would still be nice for this to get fixed though.

I think [SSH](https://docs.microsoft.com/en-us/azure/devops/repos/git/use-ssh-keys-to-authenticate?view=vsts) is probably still best to use at this moment. Check this [reply](102#issuecomment-440824208) as well. You may pass the exception, if you haven't installed _GNOME Keyring_, for example (as in #97), but then you still need to authenticate and this is another issue (see #102).

**Comment by tinycrimsonpandas** [comment]
This is still a problem under Windows Subsystem for Linux with Ubuntu 20.04 and git-credential-manager installed using homebrew.

I get the following error the first time I try to pull a repo:
```
Fatal: java.lang.RuntimeException encountered.  Details:
Secure credential storage is not available on this operating system. You may opt-in to store credentials in an unencrypted file under your user home directory by running 'git config --global credential.canFallBackToInsecureStore true'.
fatal: credential helper '!/home/linuxbrew/.linuxbrew/Cellar/openjdk/14.0.1/libexec/bin/java -Ddebug=false -Djava.net.useSystemProxies=true -jar /home/linuxbrew/.linuxbrew/Cellar/git-credential-manager/2.0.4_2/libexec/git-credential-manager-2.0.4.jar' told us to quit
```
